webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala

```diff
@@ -543,11 +543,17 @@ class FileBasedNodeConfigurationHashRepository(path: String) extends NodeConfigu
   }
 
   override def save(hashes: Set[NodeConfigurationHash]): Box[Set[NodeId]] = {
-    timeLog(s => s"Updating node configuration hashes took ${s}")(semaphore.withPermits(1) {
-      for {
-        _       <- nonAtomicWrite(NodeConfigurationHashes(hashes.toList.sortBy(_.id.value)))
-      } yield hashes.map(_.id)
-    }).toBox
+    val nodeIds = hashes.map(_.id)
+    timeLog(s => s"Updating node configuration hashes took ${s}")(
+      if(hashes.size == 0) nodeIds.succeed
+      else semaphore.withPermits(1) {
+        for {
+          current  <- nonAtomicRead()
+          filtered =  current.hashes.filterNot(h => nodeIds.contains(h.id))
```

Comment:
i'm unsure tht you need this one: you are traversing current.hashes , and n times the nodeIds

doing simply (filtered ++ hashes).toList.sortBy(_.id.value) would yield the same result

Reply:
no, `filtered` is a list, not a map, and `hashes` is not a map either, so I will have several node config hashes for the same node Id.  
I would need to do `current.hashes.map(h => (h.id, h)).toMap`, same for `hashes`, so two traverse, and one more (more coslty) for the map merge.